cmake/LinkCudaLibraries.cmake

```diff
@@ -0,0 +1,33 @@
+# Link CUDA libraries to the given target, i.e.: `target_link_libraries(target <args>)`
+#
+# Additionally makes sure CUDA stub libs don't end up being in RPath
+#
+# Example: link_cuda_libraries(mytarget PRIVATE ${CUDA_LIBRARIES})
+function(link_cuda_libraries target)
+  set(libs ${ARGN})
+  set(install_rpath "$ORIGIN")
+  set(filtered FALSE)
+  foreach(lib IN LISTS libs)
+    # CUDA stub libs are in form /prefix/lib/stubs/libcuda.so
+    # So extract the name of the parent folder, to check against "stubs"
+    # And the parent path which we need to add to the INSTALL_RPATH for non-stubs
+    get_filename_component(parent_path "${lib}" DIRECTORY)
+    get_filename_component(parent_name "${parent_path}" NAME)
+    if(parent_name STREQUAL "stubs")
+      message(STATUS "Filtering ${lib} from being set in ${target}'s RPATH, "
+                     "because it appears to point to the CUDA stubs directory.")
+      set(filtered TRUE)
+    elseif(parent_path)
+      list(APPEND install_rpath ${parent_path})
+    endif()
+  endforeach()
+
+  # Regular link command
+  target_link_libraries(${target} ${libs})
+  # Manually set INSTALL_RPATH when there were any stub libs
+  if(filtered)
+    list(REMOVE_DUPLICATES install_rpath)
+    set_target_properties(${target} PROPERTIES INSTALL_RPATH_USE_LINK_PATH FALSE)
+    set_target_properties(${target} PROPERTIES INSTALL_RPATH "${install_rpath}")
```

Comment:
This will render binary unportable if called using CUDA libraries form /usr/local/cuda-12.1`, but installed using CUDA dependencies from PYPI, wouldn't it?

Reply:
This is only executed when any stub libs are used during build, i.e. linked above. So it should not be an issue for the case you mentioned.

We need it because https://github.com/pytorch/pytorch/blob/d1ae5efa94737aea0a2944a51381e8c18f90a40e/cmake/Dependencies.cmake#L13-L16 sets `INSTALL_RPATH_USE_LINK_PATH=TRUE`

But yes this could potentially cause trouble as this sets `INSTALL_RPATH` manually to the directories of libraries passed to the **last** call to this function on this target

It hence ignores rpaths for libraries linked via CMake targets instead of paths, and those linked earlier by this function or the regular `target_link_libraries`. Not sure this is a problem in practice or if that `INSTALL_RPATH_USE_LINK_PATH` logic is useful in general at all. E.g. for the PYPI package: Why would you want/assume the users libraries are at the same location as during the build?   
However this has worked for us for a long time now.